README.md
# Flight Service Microservice Package
> by dottics PTY (LTD)

`flightserv` is a microservice package (msp) for a Go application to integrate
with the Dottics Flight Service.


CHANGELOG.md
## Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
## [0.5.0] - 2022-12-08
### Added
- The service method `DeleteFlightLog` which deletes a specific flight log for
a user.

## [0.4.0] - 2022-12-08
### Added
- The service method `UpdateFlightLog` which updates a specific flight log in
the microservice and returns updated flight log.

## [0.3.0] - 2022-12-08
### Added
- The service method `CreateFlightLog` which creates a new flight log in the
microservice and returns the new flight log.

## [0.2.0] - 2022-12-06
### Added
- Added the service method `GetFlightLog` which fetches a specific flight log
related to a specific user for a specific flight log.

## [0.1.0] - 2022-12-05
### Added
- The `AircraftType` type.
- The `GetAircraftTypes` method to get all the available aircraft types.
- The `FlightLog` type.
- Added the service method `GetFlightLogs` which fetches all the flight logs for
a user.

## [0.0.0] - 2022-12-03
### Added
- Initial commit.
- Setup layout and structure. Added `README` and `CHANGELOG`.

## [Released]


util_test.go
package flightserv

import (
	"fmt"
	"testing"
)

func TestParseBody(t *testing.T) {
	tt := []struct {
		name   string
		input  string
		output string
	}{
		{
			name:   "empty object",
			input:  `{}`,
			output: `{}`,
		},
		{
			name:   "single line",
			input:  `{ "name": "james", "age": 23 }`,
			output: `{"name":"james","age":23}`,
		},
		{
			name: "multi line",
			input: `{ 
				"name": "james", 
				"age": 23 
			}`,
			output: `{"name":"james","age":23}`,
		},
		{
			name: "an array",
			input: `{ 
				"name": "james", 
				"age": 23,
				"favNumbers": [ 1, 4, "ten" ]
			}`,
			output: `{"name":"james","age":23,"favNumbers":[1,4,"ten"]}`,
		},
	}

	for i, tc := range tt {
		name := fmt.Sprintf("%d %s", i, tc.name)
		t.Run(name, func(t *testing.T) {
			o := ParseBody(tc.input)
			if o != tc.output {
				t.Errorf("expected string '%s' got '%s'", tc.output, o)
			}
		})
	}
}


service_test.go
package flightserv

import (
	"os"
	"testing"
)

func TestNewService(t *testing.T) {
	// set the env vars NewService should automatically get them
	schemeEnv := "FLIGHT_SERVICE_SCHEME"
	scheme := "https"
	err := os.Setenv(schemeEnv, scheme)
	if err != nil {
		t.Errorf("unexpected error setting %s: %v", schemeEnv, err)
	}
	hostEnv := "FLIGHT_SERVICE_HOST"
	host := "flight.dottics.com"
	err = os.Setenv(hostEnv, host)
	if err != nil {
		t.Errorf("unexpected error setting %s: %v", hostEnv, err)
	}
	token := "my-test-token"
	ms := NewService(token)
	if ms.serv.URL.Scheme != scheme {
		t.Errorf("expected Bank Service to have Scheme %s got %s",
			scheme, ms.serv.URL.Scheme,
		)
	}
	if ms.serv.URL.Host != host {
		t.Errorf("expected Bank Service to have Host %s got %s",
			host, ms.serv.URL.Host,
		)
	}
}


flightlog_test.go
package flightserv

import (
	"fmt"
	"github.com/dottics/dutil"
	"github.com/google/uuid"
	"github.com/johannesscr/micro/microtest"
	"testing"
	"time"
)

func buildFlightLog() FlightLog {
	return FlightLog{
		UUID:              uuid.MustParse("b9782d74-dd43-484a-8fb0-f9eee8e8bcbb"),
		UserUUID:          uuid.MustParse("910b1072-b999-4a6e-b592-ef110ac5b0eb"),
		AircraftTypeUUID:  uuid.UUID{},
		AircraftType:      "A320",
		Date:              time.Date(2022, time.April, 20, 15, 35, 0, 0, time.UTC),
		Registration:      "",
		PilotInCommand:    "SELF",
		Details:           "HKG-CPT",
		InstrumentNavAids: "",
		InstrumentPlace:   "",
		InstrumentActual:  "",
		InstrumentFSTD:    1,
		InstructorSE:      0,
		InstructorME:      0,
		InstructorFSTD:    0,
		FSTD:              2,
		EngineType:        "",
		DayType:           "",
		Dual:              0,
		PIC:               1.2,
		PICUS:             0,
		Copilot:           3.3,
		DayLandings:       1,
		NightLandings:     0,
		Remarks:           "",
	}
}

func TestService_GetFlightLogs(t *testing.T) {
	tt := []struct {
		name       string
		exchange   *microtest.Exchange
		flightLogs FlightLogs
		e          dutil.Error
	}{
		{
			name: "401",
			exchange: &microtest.Exchange{
				Response: microtest.Response{
					Status: 401,
					Body: `{
						"message": "Forbidden",
						"data": {},
						"errors": {
							"permission": ["please ensure you have permission"]
						}
					}`,
				},
			},
			flightLogs: FlightLogs{},
			e: &dutil.Err{
				Status: 401,
				Errors: map[string][]string{
					"permission": {"please ensure you have permission"},
				},
			},
		},
		{
			name: "get flight logs",
			exchange: &microtest.Exchange{
				Response: microtest.Response{
					Status: 200,
					Body: `{
						"message": "flight logs found",
						"data": {
							"flightLogs": [
								{
									"uuid": "5dd41ccd-0f83-4379-b1d6-deeb26c3f281",
									"userUuid": "1d50a79a-1651-4e31-afd6-a2ce7d187a9d"
								},
								{
									"uuid": "07507951-8492-4449-95a6-3999594c5f64",
									"userUuid": "1d50a79a-1651-4e31-afd6-a2ce7d187a9d"
								},
								{
									"uuid": "a6900fa8-04fb-4de2-9a8a-3c174f3f6315",
									"userUuid": "1d50a79a-1651-4e31-afd6-a2ce7d187a9d"
								}
							]
						},
						"errors": {}
					}`,
				},
			},
			flightLogs: FlightLogs{
				{
					UUID:     uuid.MustParse("5dd41ccd-0f83-4379-b1d6-deeb26c3f281"),
					UserUUID: uuid.MustParse("1d50a79a-1651-4e31-afd6-a2ce7d187a9d"),
				},
				{
					UUID:     uuid.MustParse("07507951-8492-4449-95a6-3999594c5f64"),
					UserUUID: uuid.MustParse("1d50a79a-1651-4e31-afd6-a2ce7d187a9d"),
				},
				{
					UUID:     uuid.MustParse("a6900fa8-04fb-4de2-9a8a-3c174f3f6315"),
					UserUUID: uuid.MustParse("1d50a79a-1651-4e31-afd6-a2ce7d187a9d"),
				},
			},
			e: nil,
		},
	}

	s := NewService("")
	ms := microtest.MockServer(s.serv)

	for i, tc := range tt {
		userUUID := uuid.MustParse("1d50a79a-1651-4e31-afd6-a2ce7d187a9d")
		name := fmt.Sprintf("%d %s", i, tc.name)
		t.Run(name, func(t *testing.T) {
			ms.Append(tc.exchange)
			xfl, e := s.GetFlightLogs(userUUID)
			// test the error
			if !dutil.ErrorEqual(tc.e, e) {
				t.Errorf("expected error %v got %v", tc.e, e)
			}
			// test the query params are formatted correctly
			testURLQuery := fmt.Sprintf("userUUID=%s", userUUID.String())
			if tc.exchange.Request.URL.RawQuery != testURLQuery {
				t.Errorf("expected query '%s' got '%s'",
					testURLQuery, tc.exchange.Request.URL.RawQuery,
				)
			}
			// test the flight logs
			if !tc.flightLogs.EqualTo(xfl) {
				t.Errorf("expected flight logs %v got %v", tc.flightLogs, xfl)
			}
		})
	}
}

func TestService_GetFlightLog(t *testing.T) {
	tt := []struct {
		name      string
		exchange  *microtest.Exchange
		flightLog FlightLog
		e         dutil.Error
	}{
		{
			name: "401 Forbidden",
			exchange: &microtest.Exchange{
				Response: microtest.Response{
					Status: 401,
					Body: `{
						"message": "Forbidden",
						"data": {},
						"errors": {
							"permission": ["Please ensure you have permission"]
						}
					}`,
				},
			},
			flightLog: FlightLog{},
			e: &dutil.Err{
				Status: 401,
				Errors: map[string][]string{
					"permission": {"Please ensure you have permission"},
				},
			},
		},
		{
			name: "200 success",
			exchange: &microtest.Exchange{
				Response: microtest.Response{
					Status: 200,
					Body: `{
						"message": "flight log found",
						"data": {
							"flightLog": {
								"uuid": "b9782d74-dd43-484a-8fb0-f9eee8e8bcbb",
								"userUuid": "910b1072-b999-4a6e-b592-ef110ac5b0eb",
								"aircraftType": "A320",
								"date": "2022-04-20T15:35:00.000Z",
								"pilotInCommand": "SELF",
								"details": "HKG-CPT",
								"instrumentFSTD": 1,
								"fstd": 2,
								"pic": 1.2,
								"copilot": 3.3,
								"dayLandings": 1
							}
						}
					}`,
				},
			},
			flightLog: buildFlightLog(),
			e:         nil,
		},
	}

	s := NewService("")
	ms := microtest.MockServer(s.serv)

	testUserUUID := uuid.MustParse("71b2717c-f6aa-43f0-baf1-0fec660cfa16")
	testUUID := uuid.MustParse("905bae9d-3348-4f31-bdaf-ee7059050f63")

	for i, tc := range tt {
		name := fmt.Sprintf("%d %s", i, tc.name)
		t.Run(name, func(t *testing.T) {
			ms.Append(tc.exchange)

			f, e := s.GetFlightLog(testUserUUID, testUUID)
			// test error
			if !dutil.ErrorEqual(tc.e, e) {
				t.Errorf("expected error %v got %v", tc.e, e)
			}
			// test requests
			testQueryParams := fmt.Sprintf("UUID=%s&userUUID=%s", testUUID.String(), testUserUUID.String())
			reqQueryParams := tc.exchange.Request.URL.RawQuery
			if reqQueryParams != testQueryParams {
				t.Errorf("expected query params %v got %v", testQueryParams, reqQueryParams)
			}
			// test flight log
			if f != tc.flightLog {
				t.Errorf("expected flight log %v got %v", tc.flightLog, f)
			}
		})
	}
}

func TestService_CreateFlightLog(t *testing.T) {
	tt := []struct {
		name     string
		exchange *microtest.Exchange
		log      FlightLog
		e        dutil.Error
	}{
		{
			name: "401 Forbidden",
			exchange: &microtest.Exchange{
				Response: microtest.Response{
					Status: 401,
					Body: `{
						"message": "Forbidden",
						"data": {},
						"errors": {
							"permission": ["Please ensure you have permission"]
						}
					}`,
				},
			},
			log: FlightLog{},
			e: &dutil.Err{
				Status: 401,
				Errors: map[string][]string{
					"permission": {"Please ensure you have permission"},
				},
			},
		},
		{
			name: "201 created",
			exchange: &microtest.Exchange{
				Response: microtest.Response{
					Status: 201,
					Body: `{
						"message": "flight log created",
						"data": {
							"flightLog": {
								"uuid": "b9782d74-dd43-484a-8fb0-f9eee8e8bcbb",
								"userUuid": "910b1072-b999-4a6e-b592-ef110ac5b0eb",
								"aircraftType": "A320",
								"date": "2022-04-20T15:35:00.000Z",
								"pilotInCommand": "SELF",
								"details": "HKG-CPT",
								"instrumentFSTD": 1,
								"fstd": 2,
								"pic": 1.2,
								"copilot": 3.3,
								"dayLandings": 1
							}
						}
					}`,
				},
			},
			log: buildFlightLog(),
			e:   nil,
		},
	}

	s := NewService("")
	ms := microtest.MockServer(s.serv)

	for i, tc := range tt {
		name := fmt.Sprintf("%d %s", i, tc.name)
		t.Run(name, func(t *testing.T) {
			ms.Append(tc.exchange)

			log, e := s.CreateFlightLog(tc.log)
			if tc.exchange.Request.Method != "POST" {
				t.Errorf("expected method POST got %s", tc.exchange.Request.Method)
			}
			if !dutil.ErrorEqual(tc.e, e) {
				t.Errorf("expected error %v got %v", tc.e, e)
			}
			if log != tc.log {
				t.Errorf("expected flight log %v got %v", tc.log, log)
			}
		})
	}
}

func TestService_UpdateFlightLog(t *testing.T) {
	tt := []struct {
		name     string
		exchange *microtest.Exchange
		log      FlightLog
		e        dutil.Error
	}{
		{
			name: "401 Forbidden",
			exchange: &microtest.Exchange{
				Response: microtest.Response{
					Status: 401,
					Body: `{
						"message": "Forbidden",
						"data": {},
						"errors": {
							"permission": ["Please ensure you have permission"]
						}
					}`,
				},
			},
			log: FlightLog{},
			e: &dutil.Err{
				Status: 401,
				Errors: map[string][]string{
					"permission": {"Please ensure you have permission"},
				},
			},
		},
		{
			name: "200 updated",
			exchange: &microtest.Exchange{
				Response: microtest.Response{
					Status: 200,
					Body: `{
						"message": "flight log updated",
						"data": {
							"flightLog": {
								"uuid": "b9782d74-dd43-484a-8fb0-f9eee8e8bcbb",
								"userUuid": "910b1072-b999-4a6e-b592-ef110ac5b0eb",
								"aircraftType": "A320",
								"date": "2022-04-20T15:35:00.000Z",
								"pilotInCommand": "SELF",
								"details": "HKG-CPT",
								"instrumentFSTD": 1,
								"fstd": 2,
								"pic": 1.2,
								"copilot": 3.3,
								"dayLandings": 1
							}
						}
					}`,
				},
			},
			log: buildFlightLog(),
			e:   nil,
		},
	}

	s := NewService("")
	ms := microtest.MockServer(s.serv)

	for i, tc := range tt {
		name := fmt.Sprintf("%d %s", i, tc.name)
		t.Run(name, func(t *testing.T) {
			ms.Append(tc.exchange)

			log, e := s.UpdateFlightLog(tc.log)

			if tc.exchange.Request.Method != "PUT" {
				t.Errorf("expected method PUT got %s", tc.exchange.Request.Method)
			}
			if !dutil.ErrorEqual(tc.e, e) {
				t.Errorf("expected error %v got %v", tc.e, e)
			}
			if log != tc.log {
				t.Errorf("expected flight log %v got %v", tc.log, log)
			}
		})
	}
}

func TestService_DeleteFlightLog(t *testing.T) {
	tt := []struct {
		name     string
		exchange *microtest.Exchange
		e        dutil.Error
	}{
		{
			name: "401 Forbidden",
			exchange: &microtest.Exchange{
				Response: microtest.Response{
					Status: 401,
					Body: `{
						"message": "Forbidden",
						"errors": {
							"permission": ["Please ensure you have permission"]
						}
					}`,
				},
			},
			e: &dutil.Err{
				Status: 401,
				Errors: map[string][]string{
					"permission": {"Please ensure you have permission"},
				},
			},
		},
		{
			name: "200 success",
			exchange: &microtest.Exchange{
				Response: microtest.Response{
					Status: 200,
					Body: `{
						"message": "flight log deleted"
					}`,
				},
			},
			e: nil,
		},
	}

	s := NewService("")
	ms := microtest.MockServer(s.serv)

	testUserUUID := uuid.MustParse("71b2717c-f6aa-43f0-baf1-0fec660cfa16")
	testUUID := uuid.MustParse("905bae9d-3348-4f31-bdaf-ee7059050f63")

	for i, tc := range tt {
		name := fmt.Sprintf("%d %s", i, tc.name)
		t.Run(name, func(t *testing.T) {
			ms.Append(tc.exchange)

			e := s.DeleteFlightLog(testUserUUID, testUUID)
			// test error
			if !dutil.ErrorEqual(tc.e, e) {
				t.Errorf("expected error %v got %v", tc.e, e)
			}
			if tc.exchange.Request.Method != "DELETE" {
				t.Errorf("expected request method DELETE got %s", tc.exchange.Request.Method)
			}
			// test requests
			testQueryParams := fmt.Sprintf("UUID=%s&userUUID=%s", testUUID.String(), testUserUUID.String())
			reqQueryParams := tc.exchange.Request.URL.RawQuery
			if reqQueryParams != testQueryParams {
				t.Errorf("expected query params %v got %v", testQueryParams, reqQueryParams)
			}
		})
	}
}


type.go
package flightserv

import (
	"github.com/google/uuid"
	"time"
)

type AircraftType struct {
	UUID        uuid.UUID `json:"uuid"`
	Name        string    `json:"name"`
	Description string    `json:"description"`
}

type AircraftTypes []AircraftType

// EqualTo returns a boolean whether a slice of aircraft types are equal.
// Equal in length and equal in order of the aircraft types in the slice.
func (xa *AircraftTypes) EqualTo(aircraftTypes AircraftTypes) bool {
	if len(*xa) != len(aircraftTypes) {
		return false
	}
	for i, a := range *xa {
		aircraftType := aircraftTypes[i]
		if aircraftType != a {
			return false
		}
	}
	return true
}

type FlightLog struct {
	UUID              uuid.UUID `json:"uuid"`
	UserUUID          uuid.UUID `json:"userUuid"`
	AircraftTypeUUID  uuid.UUID `json:"aircraftTypeUuid"`
	AircraftType      string    `json:"aircraftType"`
	Date              time.Time `json:"date"`
	Registration      string    `json:"registration"`
	PilotInCommand    string    `json:"pilotInCommand"`
	Details           string    `json:"details"`
	InstrumentNavAids string    `json:"instrumentNavAids"`
	InstrumentPlace   string    `json:"instrumentPlace"`
	InstrumentActual  string    `json:"instrumentActual"`
	InstrumentFSTD    float64   `json:"instrumentFSTD"`
	InstructorSE      float64   `json:"instructorSE"`
	InstructorME      float64   `json:"instructorME"`
	InstructorFSTD    float64   `json:"instructorFSTD"`
	FSTD              float64   `json:"fstd"`
	EngineType        string    `json:"engineType"`
	DayType           string    `json:"dayType"`
	Dual              float64   `json:"dual"`
	PIC               float64   `json:"pic"`
	PICUS             float64   `json:"picus"`
	Copilot           float64   `json:"copilot"`
	DayLandings       int       `json:"dayLandings"`
	NightLandings     int       `json:"nightLandings"`
	Remarks           string    `json:"remarks"`
}

type FlightLogs []FlightLog

// EqualTo returns a boolean whether a slice of flight logs are equal.
// Equal in length and equal in order of the flight logs in the slice.
func (xf *FlightLogs) EqualTo(flightLogs FlightLogs) bool {
	if len(*xf) != len(flightLogs) {
		return false
	}
	for i, v := range *xf {
		flightLog := flightLogs[i]
		if flightLog != v {
			return false
		}
	}
	return true
}


util.go
package flightserv

import "strings"

func ParseBody(s string) string {
	// split by new line
	xs := strings.Split(s, "\n")
	ns := make([]string, 0)
	for _, si := range xs {
		ns = append(ns, strings.TrimSpace(si))
	}
	s = strings.Join(ns, "")
	// split by colon
	xs = strings.Split(s, ":")
	ns = make([]string, 0)
	for _, si := range xs {
		ns = append(ns, strings.TrimSpace(si))
	}
	s = strings.Join(ns, ":")
	// split by comma
	xs = strings.Split(s, ",")
	ns = make([]string, 0)
	for _, si := range xs {
		ns = append(ns, strings.TrimSpace(si))
	}
	s = strings.Join(ns, ",")
	// split by opening curly brace
	xs = strings.Split(s, "{")
	ns = make([]string, 0)
	for _, si := range xs {
		ns = append(ns, strings.TrimSpace(si))
	}
	s = strings.Join(ns, "{")
	// split by closing curly brace
	xs = strings.Split(s, "}")
	ns = make([]string, 0)
	for _, si := range xs {
		ns = append(ns, strings.TrimSpace(si))
	}
	s = strings.Join(ns, "}")
	// split by opening square brace
	xs = strings.Split(s, "[")
	ns = make([]string, 0)
	for _, si := range xs {
		ns = append(ns, strings.TrimSpace(si))
	}
	s = strings.Join(ns, "[")
	// split by closing square brace
	xs = strings.Split(s, "]")
	ns = make([]string, 0)
	for _, si := range xs {
		ns = append(ns, strings.TrimSpace(si))
	}
	s = strings.Join(ns, "]")
	return s
}


service.go
package flightserv

import "github.com/johannesscr/micro/msp"

// Service is the type that represents the communication with a microservice.
type Service struct {
	serv *msp.Service
}

const microServiceName string = "flight"

// NewService creates a microservice-package (msp) instance. The msp
// is an instance that loads the environmental variables to be able
// to connect to the specific microservice. The msp contains all the
// implementations to correctly exchange with the microservice.
//
// Once a user has logged in the user receives a token, that token needs
// to be passed to the new service so that the token can be added to headers
// to gain access to the microservice.
func NewService(token string) *Service {
	s := &Service{
		serv: msp.NewService(token, microServiceName),
	}
	return s
}


aircraft_type.go
package flightserv

import "github.com/dottics/dutil"

// GetAircraftTypes gets all the aircraft types from the flight service. It
// returns all the aircraft types available.
func (s *Service) GetAircraftTypes() (AircraftTypes, dutil.Error) {
	// set path
	s.serv.URL.Path = "/aircraft-type"
	// do request
	r, e := s.serv.NewRequest("GET", s.serv.URL.String(), nil, nil)
	if e != nil {
		return AircraftTypes{}, e
	}

	// response structure
	type Data struct {
		AircraftTypes `json:"aircraftTypes"`
	}
	res := struct {
		Data   `json:"data"`
		Errors map[string][]string `json:"errors"`
	}{}

	// decode the response
	_, e = s.serv.Decode(r, &res)
	if e != nil {
		return AircraftTypes{}, e
	}

	if r.StatusCode != 200 {
		e := &dutil.Err{
			Status: r.StatusCode,
			Errors: res.Errors,
		}
		return AircraftTypes{}, e
	}
	// return aircraft types on success
	return res.Data.AircraftTypes, nil
}


flightlog.go
package flightserv

import (
	"github.com/dottics/dutil"
	"github.com/google/uuid"
	"net/url"
)

// GetFlightLogs gets all the flight logs for a specific user.
func (s *Service) GetFlightLogs(UserUUID uuid.UUID) (FlightLogs, dutil.Error) {
	// set path
	s.serv.URL.Path = "log"
	// set query params
	qs := url.Values{"userUUID": {UserUUID.String()}}
	s.serv.URL.RawQuery = qs.Encode()
	// do request
	r, e := s.serv.NewRequest("GET", s.serv.URL.String(), nil, nil)
	if e != nil {
		return FlightLogs{}, e
	}

	// response structure
	type Data struct {
		FlightLogs `json:"flightLogs"`
	}
	res := struct {
		Data   `json:"data"`
		Errors dutil.Errors `json:"errors"`
	}{}
	// decode the response
	_, e = s.serv.Decode(r, &res)
	if e != nil {
		return FlightLogs{}, e
	}

	if r.StatusCode != 200 {
		e := &dutil.Err{
			Status: r.StatusCode,
			Errors: res.Errors,
		}
		return FlightLogs{}, e
	}
	return res.Data.FlightLogs, nil
}

func (s *Service) GetFlightLog(userUUID, UUID uuid.UUID) (FlightLog, dutil.Error) {
	// set path
	s.serv.URL.Path = "/log/-"
	// set query params
	qs := url.Values{"userUUID": {userUUID.String()}, "UUID": {UUID.String()}}
	s.serv.URL.RawQuery = qs.Encode()
	// do request
	r, e := s.serv.NewRequest("GET", s.serv.URL.String(), nil, nil)
	if e != nil {
		return FlightLog{}, e
	}

	// response structure
	type Data struct {
		FlightLog `json:"flightLog"`
	}
	res := struct {
		Data   `json:"data"`
		Errors dutil.Errors `json:"errors"`
	}{}
	// decode the response
	_, e = s.serv.Decode(r, &res)
	if e != nil {
		return FlightLog{}, e
	}

	if r.StatusCode != 200 {
		e := &dutil.Err{
			Status: r.StatusCode,
			Errors: res.Errors,
		}
		return FlightLog{}, e
	}
	return res.Data.FlightLog, nil
}

// CreateFlightLog creates a new FlightLog entry.
func (s *Service) CreateFlightLog(log FlightLog) (FlightLog, dutil.Error) {
	// set path
	s.serv.URL.Path = "/log"
	// parse body
	p, e := dutil.MarshalReader(log)
	if e != nil {
		return FlightLog{}, e
	}
	// do request
	r, e := s.serv.NewRequest("POST", s.serv.URL.String(), nil, p)
	if e != nil {
		return FlightLog{}, e
	}
	// response structure
	type Data struct {
		FlightLog `json:"flightLog"`
	}
	res := struct {
		Data         `json:"data"`
		dutil.Errors `json:"errors"`
	}{}
	// decode response
	_, e = s.serv.Decode(r, &res)
	if e != nil {
		return FlightLog{}, e
	}

	if r.StatusCode != 201 {
		e := &dutil.Err{
			Status: r.StatusCode,
			Errors: res.Errors,
		}
		return FlightLog{}, e
	}

	return res.Data.FlightLog, nil
}

// UpdateFlightLog updates a specific FlightLog entry.
func (s *Service) UpdateFlightLog(log FlightLog) (FlightLog, dutil.Error) {
	// set path
	s.serv.URL.Path = "/log/-"
	// parse body
	p, e := dutil.MarshalReader(log)
	if e != nil {
		return FlightLog{}, e
	}
	// do request
	r, e := s.serv.NewRequest("PUT", s.serv.URL.String(), nil, p)
	if e != nil {
		return FlightLog{}, e
	}
	// response structure
	type Data struct {
		FlightLog `json:"flightLog"`
	}
	res := struct {
		Data         `json:"data"`
		dutil.Errors `json:"errors"`
	}{}
	// decode response
	_, e = s.serv.Decode(r, &res)
	if e != nil {
		return FlightLog{}, e
	}

	if r.StatusCode != 200 {
		e := &dutil.Err{
			Status: r.StatusCode,
			Errors: res.Errors,
		}
		return FlightLog{}, e
	}

	return res.Data.FlightLog, nil
}

// DeleteFlightLog deletes a specific flight log for a user from the
// microservice and only returns an error if the is one. If nil is returned
// then the FlightLog was deleted successfully.
func (s *Service) DeleteFlightLog(userUUID, UUID uuid.UUID) dutil.Error {
	// set path
	s.serv.URL.Path = "/log/-"
	// set query params
	qs := url.Values{"userUUID": {userUUID.String()}, "UUID": {UUID.String()}}
	s.serv.URL.RawQuery = qs.Encode()
	// do request
	r, e := s.serv.NewRequest("DELETE", s.serv.URL.String(), nil, nil)
	if e != nil {
		return e
	}

	res := struct {
		Message string       `json:"message"`
		Errors  dutil.Errors `json:"errors"`
	}{}
	// decode the response
	_, e = s.serv.Decode(r, &res)
	if e != nil {
		return e
	}

	if r.StatusCode != 200 {
		e := &dutil.Err{
			Status: r.StatusCode,
			Errors: res.Errors,
		}
		return e
	}
	return nil
}
